### PR TITLE
Added new target support to Boolean, String, Number Object

### PR DESCRIPTION
**Pull request by kisbg**
JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 1]
```diff
@@ -566,12 +576,39 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
...
+  return ecma_op_create_class_object (proto_id, value, lit_id);
 } /* ecma_op_to_object */
 
+/**
+ * create class object operation.
+ *
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value
+ */
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-conversion.c:595]:
```c
/**
 * Create a ECMA_OBJECT_TYPE_CLASS object from the given arguments.
 *
 * @return ecma_value - constructed object
 */
```

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 2]
```diff
@@ -566,12 +576,39 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
...
+ecma_value_t
+ecma_op_create_class_object (ecma_builtin_id_t proto_id, /**< prototype id */
+                             ecma_value_t value, /**< ecma value */
+                             uint16_t lit_id) /**< magic string id */
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-conversion.c:599]: `lit_id` ->`class_id`

File: jerry-core/ecma/operations/ecma-number-object.c [hunk 3]
```diff
@@ -43,19 +45,30 @@ ecma_op_create_number_object (ecma_value_t arg) /**< argument passed to the Numb
 {
   ecma_number_t num;
   ecma_value_t conv_to_num_completion = ecma_op_to_number (arg, &num);
-
+  ecma_builtin_id_t proto_id;
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-number-object.c:48]: Lower these declarations closer to these first usage.

File: jerry-core/ecma/operations/ecma-string-object.c [hunk 4]
```diff
@@ -47,6 +49,8 @@ ecma_op_create_string_object (const ecma_value_t *arguments_list_p, /**< list of
                 || arguments_list_p != NULL);
 
   ecma_value_t prim_value = ecma_make_magic_string_value (LIT_MAGIC_STRING__EMPTY);
+  ecma_builtin_id_t proto_id;
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-string-object.c:52]: ditto.

**Review by rerobika — CHANGES_REQUESTED** [review]

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 5]
```diff
@@ -566,12 +576,38 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
...
     }
   }
+
+  return ecma_op_create_class_object (proto_id, value, lit_id);
 } /* ecma_op_to_object */
 
+/**
+ * Create a ECMA_OBJECT_TYPE_CLASS object from the given arguments.
+ *
+ *
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-conversion.c:592]: Only one new comment line.

File: jerry-core/ecma/operations/ecma-number-object.c [hunk 6]
```diff
@@ -43,19 +45,29 @@ ecma_op_create_number_object (ecma_value_t arg) /**< argument passed to the Numb
 {
   ecma_number_t num;
   ecma_value_t conv_to_num_completion = ecma_op_to_number (arg, &num);
-
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-number-object.c:46]: Do not remove this line.

File: jerry-core/ecma/operations/ecma-string-object.c [hunk 7]
```diff
@@ -47,7 +49,6 @@ ecma_op_create_string_object (const ecma_value_t *arguments_list_p, /**< list of
                 || arguments_list_p != NULL);
 
   ecma_value_t prim_value = ecma_make_magic_string_value (LIT_MAGIC_STRING__EMPTY);
-
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-string-object.c:50]: Ditto.

File: jerry-core/ecma/operations/ecma-string-object.c [hunk 8]
```diff
@@ -60,20 +61,36 @@ ecma_op_create_string_object (const ecma_value_t *arguments_list_p, /**< list of
...
+  }
+#endif /* ENABLED (JERRY_ESNEXT) */
   ecma_object_t *object_p = ecma_create_object (prototype_obj_p,
                                                 sizeof (ecma_extended_object_t),
                                                 ECMA_OBJECT_TYPE_CLASS);
 
   ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) object_p;
   ext_object_p->u.class_prop.class_id = LIT_MAGIC_STRING_STRING_UL;
   ext_object_p->u.class_prop.u.value = prim_value;
-
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-string-object.c:76]: Ditto.

**Review by rerobika — CHANGES_REQUESTED** [review]

**Review by rerobika — APPROVED** [review]
LGTM

File: jerry-core/ecma/operations/ecma-boolean-object.c [hunk 9]
```diff
@@ -42,13 +44,27 @@ ecma_value_t
...
+  proto_id = ECMA_BUILTIN_ID_BOOLEAN_PROTOTYPE;
 #else /* ENABLED (JERRY_BUILTIN_BOOLEAN) */
-  ecma_object_t *prototype_obj_p = ecma_builtin_get (ECMA_BUILTIN_ID_OBJECT_PROTOTYPE);
+  proto_id = ECMA_BUILTIN_ID_OBJECT_PROTOTYPE;
 #endif /* !(ENABLED (JERRY_BUILTIN_BOOLEAN) */
 
+  ecma_object_t *prototype_obj_p = ecma_builtin_get (proto_id);
+
+#if ENABLED (JERRY_ESNEXT)
+  if (JERRY_CONTEXT (current_new_target))
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-boolean-object.c:58]: I would copy this into a local variable.

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 10]
```diff
@@ -35,6 +35,7 @@
 #include "ecma-string-object.h"
 #include "ecma-symbol-object.h"
 #include "jrt-libc-includes.h"
+#include "jcontext.h"
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-conversion.c:38]: Why is this needed?

File: jerry-core/ecma/operations/ecma-boolean-object.c [hunk 11]
```diff
@@ -42,13 +44,27 @@ ecma_value_t
 ecma_op_create_boolean_object (ecma_value_t arg) /**< argument passed to the Boolean constructor */
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-boolean-object.c:44]: An important question: is this only used by constructors? Because if it is used elsewhere, it must not use the new target. This is also true for all other functions.

File: tests/test262-esnext-excludelist.xml [hunk 12]
```diff
@@ -9668,7 +9666,6 @@
   <test id="built-ins/RegExp/prototype/sticky/cross-realm.js"><reason></reason></test>
   <test id="built-ins/RegExp/prototype/unicode/cross-realm.js"><reason></reason></test>
   <test id="built-ins/SharedArrayBuffer/proto-from-ctor-realm.js"><reason></reason></test>
-  <test id="built-ins/String/proto-from-ctor-realm.js"><reason></reason></test>
```
Comment (zherczeg) [inline tests/test262-esnext-excludelist.xml:9671]: Only realm tests are fixed? No normal tests for these?

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: tests/test262-esnext-excludelist.xml (hunk 12 above)
Comment (kisbg) [reply tests/test262-esnext-excludelist.xml:9671]: Sadly, there is no new target test for boolean, string and number object.

File: jerry-core/ecma/operations/ecma-boolean-object.c (hunk 11 above)
Comment (kisbg) [reply jerry-core/ecma/operations/ecma-boolean-object.c:44]: Yes, only used `ecma_builtin_boolean_dispatch_construct`.

**Review by zherczeg — APPROVED** [review]
LGTM
